manpages/falrun.1
.\" Process this file with
.\" groff -man -Tascii falrun.1
.\"
.TH flcrun 1 "April 2007" "Falcon toolset" "Falcon User Manuals"
.SH NAME

flcrun \- The Falcon command line execution tool

.SH SYNOPSIS

.B falrun [options] [main_script] [script_options]

.SH DESCRIPTION

The
.B falrun
command line tool is a subset of the
.BR falcon(1)
command line interpreter which only executes
pre\-compiled falcon scripts in ".fam" module
files. This is meant to launch Falcon based
application which were compiled on a different
system, and whose source is not shipped.

.SH OPTIONS

.IP \-e\ <enc>
Set given encoding as default for VM I/O. Unless the scripts
select a different I/O encoding, the streams that are provided
to the falcon VM (like the output stream for printf) are encoded
using the given ISO encoding. This overrides the default encoding
that is detected by reading the environment settings. In example,
if your system uses iso\-8859\-1 encoding by default, but you want
your script to read and write utf\-8 files, use the option
.B	\-e\ utf\-8

.IP \-h\ or\ \-?
Prints a brief help on stdout and exits.
.IP \-L\ <path>
Changes the default load path. This overrides both the internal
built in settings and the contents of environment variable
FALCON_LOAD_PATH. Each directory in the path should be separated
by ";" and use forward slashes, like this:

.B		falrun \-L "./;/usr/share/falcon_mod;./myapp"

.IP \-p\ <module>
Preloads the given module as if it were loaded by the main script.
.IP \-P
Ignore default load paths and uses only the paths set via the \-L switch.
.IP \-v
Prints copyright notice and version and exits.

.SH FILES

.I /usr/lib/libfalcon_engine.so
.RS
Default location of the Falcon Engine loadable module.
.RE

.I /usr/lib/falcon
.RS
Default directory containing Falcon binary modules.
.RE

.SH ENVIRONMENT

.IP FALCON_LOAD_PATH
Default search path for modules loaded by the scripts.

.IP FALCON_SRC_ENCODING
Default encoding for the source scripts loaded by falcon (when
different from the system default).

.IP FALCON_VM_ENCODING
Default encoding for the VM I/O streams (when different
from system default).


.SH AUTHOR

Giancarlo Niccolai <gc@falconpl.org>

.SH "SEE ALSO"

.BR falcon (1)
.BR faldisass (1)

.SH LICENSE
This document is released under the "GNU Free Documentation License, version 1.2".
On Debian systems, the complete text of the Free Documentation License, version 1.2,
can be found in /usr/share/common\-licenses/.



NEWS

                    The Falcon Programming Language

                        News for release 0.8


New repository architecture and build system.



clt/falrun/falrun.cpp
/*
   FALCON - The Falcon Programming Language.
   FILE: falrun.cpp

   A simple program that uses Falcon VM to execute falcon compiled codes.
   -------------------------------------------------------------------
   Author: Giancarlo Niccolai
   Begin: dom ago  8 20:30:13 CEST 2004

   -------------------------------------------------------------------
   (C) Copyright 2004: the FALCON developers (see list in AUTHORS file)

   See LICENSE file for licensing details.
*/

#include <falcon/vm.h>
#include <falcon/sys.h>
#include <falcon/modloader.h>
#include <falcon/runtime.h>
#include <falcon/core_ext.h>
#include <falcon/string.h>
#include <falcon/carray.h>
#include <falcon/string.h>
#include <falcon/stdstreams.h>
#include <falcon/transcoding.h>
#include <falcon/fstream.h>
#include <falcon/fassert.h>
#include <falcon/genericlist.h>


using namespace Falcon;

List preloaded;

String load_path;
String module_language;
Stream *stdIn;
Stream *stdOut;
Stream *stdErr;
String io_encoding;
bool ignore_defpath = false;


static void version()
{
   stdOut->writeString( "FALCON runtime virtual machine.\n" );
   stdOut->writeString( "Version " );
   stdOut->writeString( FALCON_VERSION " (" FALCON_VERSION_NAME ")" );
   stdOut->writeString( "\n" );
   stdOut->flush();
}

static void usage()
{
   stdOut->writeString( "Usage: falrun [options] file.fam [script options]\n" );
   stdOut->writeString( "\n" );
   stdOut->writeString( "Options:\n" );
   stdOut->writeString( "   -e <enc>    select default encoding for VM streams\n" );
   stdOut->writeString( "   -h          this help\n" );
   stdOut->writeString( "   -p mod      pump one or more module in the virtual machine\n" );
   stdOut->writeString( "   -l <lang>   Set preferential language of loaded modules\n" );
   stdOut->writeString( "   -L<path>    set path for 'load' directive\n" );
   stdOut->writeString( "   -P          Ignore default system paths\n" );
   stdOut->writeString( "   -v          print copyright notice and version and exit\n" );
   stdOut->writeString( "\n" );
   stdOut->writeString( "Paths must be in falcon file name format: directory separatros must be slashes and\n" );
   stdOut->writeString( "multiple entries must be entered separed by a semicomma (';')\n" );
   stdOut->writeString( "\n" );
   stdOut->flush();
}

void findModuleName( const String &filename, String &name )
{
   uint32 pos = filename.rfind( "/" );
   if ( pos == csh::npos ) {
      name = filename;
   }
   else {
      name = filename.subString( pos + 1 );
   }

   // find the dot
   pos = name.rfind( "." );
   if ( pos != csh::npos ) {
      name = name.subString( 0, pos );
   }
}

void findModulepath( const String &filename, String &path )
{
   uint32 pos = filename.rfind( "/" );
   if ( pos != csh::npos ) {
      path = filename.subString( 0, pos );
   }
}


String get_load_path()
{
   String envpath;
   bool hasEnvPath = Sys::_getEnv ( "FALCON_LOAD_PATH", envpath );

   String lp;

   if ( load_path.size() > 0 )
   {
      if ( ignore_defpath )
         lp = load_path;
      else
         lp = FALCON_DEFAULT_LOAD_PATH ";" + load_path;
   }

   if ( ! hasEnvPath || ignore_defpath )
   {
      return lp;
   }
   else
   {
      if ( lp == "" )
         return envpath;
      else
         return lp + ";" + envpath;
   }

   return lp;  // just a warning
}


String get_io_encoding()
{
   if ( io_encoding != "" )
      return io_encoding;

   String ret;
   if ( Sys::_getEnv("FALCON_VM_ENCODING", ret) )
      return ret;

   return "";
}


int main( int argc, char *argv[] )
{
   // Install a void ctrl-c handler (let ctrl-c to kill this app)
   Sys::_dummy_ctrl_c_handler();

   Falcon::Engine::AutoInit autoInit;


   int script_pos = argc;
   char *input_file = 0;
   FileStream *bincode_stream;

   stdOut = stdOutputStream();
   stdErr = stdErrorStream();
   stdIn = stdInputStream();

   // option decoding
   for ( int i = 1; i < argc; i++ )
   {
      char *op = argv[i];

      if (op[0] == '-' )
      {
         switch ( op[1] )
         {
            case 'e':
               if ( op[2] == 0 && i < argc + 1) {
                  io_encoding = argv[++i];
               }
               else {
                  io_encoding = op + 2;
               }
            break;

            case 'h': usage(); return 0;

            case 'L':
               if ( op[2] == 0 && i < argc + 1)
                  load_path = argv[++i];
               else
                  load_path = op + 2; break;
            break;

            case 'P':
               ignore_defpath = true;
               break;

            case 'l':
               if ( op[2] == 0 && i + 1 < argc )
                  module_language = argv[++i];
               else
                  module_language = op + 2;
            break;

            case 'p':
               if ( op[2] == 0 && i < argc + 1)
                  preloaded.pushBack( argv[++i] );
               else
                  preloaded.pushBack( op + 2 );
            break;


            case 'v': version(); return 0;

            default:
               stdOut->writeString( "falrun: unrecognized option '" );
               stdOut->writeString( op );
               stdOut->writeString( "'.\n\n" );
               usage();
               return 1;
         }
      }
      else {
         input_file = op;
         script_pos = i+1;
         break;
      }
   }

   // eventually change the encodings.
   io_encoding = get_io_encoding();

   if ( io_encoding != "" )
   {
      Transcoder *trans = TranscoderFactory( io_encoding, 0, true );
      if ( trans == 0 )
      {
         stdOut->writeString( "Fatal: unrecognized encoding '" + io_encoding + "'.\n\n" );
         return 1;
      }
      delete stdIn ;
      delete stdOut;
      delete stdErr;

      trans->setUnderlying( new StdInStream );

      stdIn = AddSystemEOL( trans, true );
      stdOut = AddSystemEOL( TranscoderFactory( io_encoding, new StdOutStream, true ), true );
      stdErr = AddSystemEOL( TranscoderFactory( io_encoding, new StdErrStream, true ), true );
   }

   if ( input_file == 0 ) {
      stdOut->writeString( "falrun: missing script name.\n" );
      usage();
      return 1;
   }

   bincode_stream = new FileStream;
   bincode_stream->open( input_file );

   if ( ! bincode_stream->good() )
   {
      stdOut->writeString( "falrun: Can't open file " );
      stdOut->writeString( input_file );
      stdOut->writeString( "\n" );
      stdOut->flush();
      return 1;
   }


   String module_name;
   String source_path;
   findModuleName( input_file, module_name );
   findModulepath( input_file, source_path );

   //-----------------------------------------
   // execute the script.
   //

   if ( source_path != "" )
      source_path += ";";

   try
   {
      ModuleLoader *modloader = new ModuleLoader( source_path + get_load_path() );
      Engine::setSearchPath( modloader->getSearchPath() );

      // set the module preferred language; ok also if default ("") is used
      modloader->setLanguage( module_language );

      Module *core = core_module_init();

      Module *main_mod = modloader->loadModule( bincode_stream );

      VMachine *vmachine = new VMachine(false);
      // change default machine streams.
      vmachine->stdIn( stdIn );
      vmachine->stdOut( stdOut );
      vmachine->stdErr( stdErr );
      vmachine->init();

      vmachine->link( core );
      core->decref();
      Runtime *runtime = new Runtime( modloader );

      // preload required modules

      ListElement *pliter = preloaded.begin();
      while( pliter != 0 )
      {
         Module *module = modloader->loadName( * ((String *) pliter->data()) );
         runtime->addModule( module );
         pliter = pliter->next();
      }

      Item *item_args = vmachine->findGlobalItem( "args" );
      fassert( item_args != 0 );
      CoreArray *args = new CoreArray( argc - script_pos );
      for ( int ap = script_pos; ap < argc; ap ++ ) {
         args->append( new CoreString( argv[ap] ) );
      }
      item_args->setArray( args );

      Item *script_name = vmachine->findGlobalItem( "scriptName" );
      fassert( script_name != 0 );
      script_name->setString( new CoreString( module_name ) );

      // the runtime will try to load the references.
      runtime->addModule( main_mod );

      if( vmachine->link( runtime ) )
      {
         vmachine->launch();

         if ( vmachine->regA().type() == FLC_ITEM_INT )
            return (int32) vmachine->regA().asInteger();

         return 0;
      }

      vmachine->finalize();
   }
   catch ( Error *err )
   {
      String temp;
      err->toString( temp );
      stdErr->writeString( "falcon: FATAL - Program terminated with error.\n" );
      stdErr->writeString( temp + "\n" );
      err->decref();
      return 1;
   }


   return 255;
}


/* end of falrun.cpp */
